kernel/arch/dreamcast/hardware/maple/vmu.c

```diff
@@ -85,33 +85,33 @@ static void vmu_poll_reply(maple_state_t *st, maple_frame_t *frm) {
     if(respbuf[0] != MAPLE_FUNC_CLOCK)
         return;
 
-    /* Update the status area from the response */
-    if(frm->dev) {
-        const maple_device_t *cont = maple_enum_dev(frm->dev->port, 0);
-
-        /* Verify the size of the frame and grab a pointer to it */
-        //assert(resp->data_len == 7);
-        raw = (vmu_cond_t *)(respbuf + 1);
-
-        /* Fill the "nice" struct from the raw data */
-        cooked = (vmu_state_t *)(frm->dev->status);
-        /* Invert raw struct as nice struct */
-        cooked->buttons = ~(*raw);
-
-        /* Check to see if the VMU is upside-down in the controller and readjust
-           its directional buttons accordingly. */
-        if(cont && (cont->info.functions & MAPLE_FUNC_CONTROLLER) &&
-           (frm->dev->info.connector_direction == cont->info.connector_direction)) {
-            cooked->buttons = (cooked->buttons & 0xf0)  |
-                              (cooked->dpad_up    << 1) | /* down */
-                              (cooked->dpad_down  << 0) | /* up */
-                              (cooked->dpad_left  << 3) | /* right */
-                              (cooked->dpad_right << 2);  /* left */
+    if(!frm->dev)
+        return;
 
-        }
+    /* Verify the size of the frame and grab a pointer to it */
+    //assert(resp->data_len == 7);
```

Comment:
This line still meant to be commented out?

Reply:
Yea, I'm not familiar enough with the vmu button input stuff to have uncommented it and given it the same treatment. I'm can't be the same as the rest ( `assert(sizeof(vmu_cond_t) == ((resp->data_len - 1) * sizeof(uint32_t)));` ) because vmu_cond_t is only one byte.

This might be something @gyrovorbis could comment on, but I figured the proper thing would be to de-magic that number. Perhaps vmu_cond_t is actually meant to be 7 bytes and the rest is just dummy data, or not useful? (Like in mouse_cond_t).

Reply:
Yeah, actually that's what it looks like. The BIOS is just writing 0s for the rest of the bytes.

Reply:
So, turns out that the 7 is just outright wrong. `resp->data_len` comes back as '2' which is the minimum size. It has one uint32 for the functions mask, then one for the payload. Only the first byte of the payload is used so the rest is dummy data. This makes a lot of sense since it just doesn't have anything else to add but the units are in uint32s for the response.

I've updated the vmu_cond_t to properly have those dummy bytes, which needed a bit of adjustment. Then the same assert pattern was possible as in every other polling function.

@gyrovorbis It would be great if you could test if this still works. I just don't have any controller that has the buttons exposed, and hopefully you still have the test program you wrote for this :) . Some day I'll take a hacksaw to one of my spare controllers to be able to fiddle with this more.

Reply:
> This line still meant to be commented out? I see that it existed in the old source. Do you know if we still need it or not?

Thanks for pointing it out. It was less work than I'd expected, and it really helped standardize that driver to match the rest. AND document the hardware slightly better :)